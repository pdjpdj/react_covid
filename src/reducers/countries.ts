export interface CountryItem {
  Country: string;
  Slug: string;
  ISO2: string;
  latitude: number;
  longitude: number;
}

export interface RapidCountryItem {
  Name: string;
  alpha2code: string;
  alpha3code: string;
  latitude: number;
  longitude: number;
}

interface CountriesState {
  loading: boolean;
  error: boolean;
  countries: CountryItem[] | undefined;
}

export interface CountriesAction {
  type: string;
  payload?: CountryItem[];
}

const countriesActionTypes = {
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  FAILURE: 'FAILURE',
};

export const countriesActionCreators = {
  loading: () => ({type: countriesActionTypes.LOADING}),
  failure: () => ({type: countriesActionTypes.FAILURE}),
  success: (payload: CountryItem[]) => ({
    type: countriesActionTypes.SUCCESS,
    payload,
  }),
};

export const countriesInitialState: CountriesState = {
  loading: true,
  error: false,
  countries: undefined,
};

export function countriesReducer(
  state: CountriesState,
  action: CountriesAction,
): CountriesState {
  switch (action.type) {
    case countriesActionTypes.LOADING:
      return {...state, loading: true, error: false};
    case countriesActionTypes.FAILURE:
      return {...state, loading: false, error: true};
    case countriesActionTypes.SUCCESS:
      return {
        ...state,
        loading: false,
        error: false,
        countries: action.payload,
      };
    default:
      return state;
  }
}
