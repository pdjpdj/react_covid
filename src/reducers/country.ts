export interface CountryTotalItem {
  Country: string;
  CountryCode: string;
  Province: string;
  City: string;
  CityCode: string;
  Confirmed: number;
  Deaths: number;
  Recovered: number;
  Active: number;
  Date: string;
}

interface CountryState {
  loading: boolean;
  error: boolean;
  totals: CountryTotalItem[] | undefined;
}

export interface CountryAction {
  type: string;
  payload?: CountryTotalItem[];
}

const countryActionTypes = {
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  FAILURE: 'FAILURE',
};

export const countryActionCreators = {
  loading: () => ({type: countryActionTypes.LOADING}),
  failure: () => ({type: countryActionTypes.FAILURE}),
  success: (payload: CountryTotalItem[]) => ({
    type: countryActionTypes.SUCCESS,
    payload,
  }),
};

export const countryInitialState: CountryState = {
  loading: true,
  error: false,
  totals: [],
};

export function countryReducer(
  state: CountryState,
  action: CountryAction,
): CountryState {
  switch (action.type) {
    case countryActionTypes.LOADING:
      return {...state, loading: true, error: false};
    case countryActionTypes.FAILURE:
      return {...state, loading: false, error: true};
    case countryActionTypes.SUCCESS:
      return {...state, loading: false, error: false, totals: action.payload};
    default:
      return state;
  }
}
