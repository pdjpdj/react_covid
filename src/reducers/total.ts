export interface TotalItem {
  NewConfirmed: number;
  TotalConfirmed: number;
  NewDeaths: number;
  TotalDeaths: number;
  NewRecovered: number;
  TotalRecovered: number;
}

interface TotalState {
  loading: boolean;
  error: boolean;
  totals: TotalItem | undefined;
}

export interface TotalAction {
  type: string;
  payload?: TotalItem;
}

const totalActionTypes = {
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  FAILURE: 'FAILURE',
};

export const totalActionCreators = {
  loading: () => ({type: totalActionTypes.LOADING}),
  failure: () => ({type: totalActionTypes.FAILURE}),
  success: (payload: TotalItem) => ({type: totalActionTypes.SUCCESS, payload}),
};

export const totalInitialState: TotalState = {
  loading: true,
  error: false,
  totals: undefined,
};

export function totalReducer(
  state: TotalState,
  action: TotalAction,
): TotalState {
  switch (action.type) {
    case totalActionTypes.LOADING:
      return {...state, loading: true, error: false};
    case totalActionTypes.FAILURE:
      return {...state, loading: false, error: true};
    case totalActionTypes.SUCCESS:
      return {...state, loading: false, error: false, totals: action.payload};
    default:
      return state;
  }
}
