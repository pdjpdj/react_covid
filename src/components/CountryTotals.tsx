import React from 'react';
import {CountryTotalItem} from '../reducers/country';
import {formatMidDate} from '../utils/dateFormatter';

interface CountryTotalsProps {
  totals: CountryTotalItem[] | undefined;
}

function CountryTotals({totals}: CountryTotalsProps) {
  if (!totals || totals.length === 0) {
    return (
      <div>
        <div>
          No data for this country
        </div>
      </div>
    );
  }

  return (
    <div>
      <table>
        <caption className='totals-subheader'>
          Latest ({formatMidDate(new Date(totals[totals.length - 1].Date))})
        </caption>
        <tbody>
          <tr className='table-row'>
            <td className='table-row-label'>
              Confirmed:
            </td>
            <td className='table-row-data'>
              {(
                totals[totals.length - 1].Confirmed -
                totals[totals.length - 2].Confirmed
              ).toLocaleString()}
            </td>
          </tr>
          <tr className='table-row table-row-alt'>
            <td className='table-row-label'>
              Deaths:
            </td>
            <td className='table-row-data'>
              {(
                totals[totals.length - 1].Deaths -
                totals[totals.length - 2].Deaths
              ).toLocaleString()}
            </td>
          </tr>
          <tr className='table-row'>
            <td className='table-row-label'>
              Recovered:
            </td>
            <td className='table-row-data'>
              {(
                totals[totals.length - 1].Recovered -
                totals[totals.length - 2].Recovered
              ).toLocaleString()}
            </td>
          </tr>
        </tbody>
      </table>
      <table>
        <caption className='totals-subheader'>
          All time
        </caption>
        <tbody>
          <tr className='table-row'>
            <td className='table-row-label'>
              Confirmed:
            </td>
            <td className='table-row-data'>
              {totals[totals.length - 1].Confirmed.toLocaleString()}
            </td>
          </tr>
          <tr className='table-row table-row-alt'>
            <td className='table-row-label'>
              Deaths:
            </td>
            <td className='table-row-data'>
              {totals[totals.length - 1].Deaths.toLocaleString()}
            </td>
          </tr>
          <tr className='table-row'>
            <td className='table-row-label'>
              Recovered:
            </td>
            <td className='table-row-data'>
              {totals[totals.length - 1].Recovered.toLocaleString()}
            </td>
          </tr>
          <tr className='table-row table-row-alt'>
            <td className='table-row-label'>
              Active:
            </td>
            <td className='table-row-data'>
              {totals[totals.length - 1].Active.toLocaleString()}
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  );
}

export default CountryTotals;
