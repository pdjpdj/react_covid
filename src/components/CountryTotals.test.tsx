import React from 'react';
import { render, screen } from '@testing-library/react';
import CountryTotals from './CountryTotals';
import { CountryTotalItem } from '../reducers/country';

const mockTotals: CountryTotalItem[] = [
  {
    Country: 'United Kingdom',
    CountryCode: 'UK',
    Province: '',
    City: '',
    CityCode: '',
    Confirmed: 4500,
    Deaths: 123,
    Recovered: 340,
    Active: 4160,
    Date: Date().toString()
  },
  {
    Country: 'United Kingdom',
    CountryCode: 'UK',
    Province: '',
    City: '',
    CityCode: '',
    Confirmed: 9000,
    Deaths: 256,
    Recovered: 680,
    Active: 8320,
    Date: Date().toString()
  }
]

test('renders country totals with data matches snapshot', () => {
  const {container} = render(<CountryTotals totals={mockTotals} />);
  expect(container.firstChild).toMatchSnapshot();
});

test('renders no data when no table data', () => {
  render(<CountryTotals totals={[]} />);
  const text = screen.getAllByText(/^No data for this country$/);
  expect(text.length).toEqual(1);
});

test('renders no data when no table undefined', () => {
  render(<CountryTotals totals={undefined} />);
  const text = screen.getAllByText(/^No data for this country$/);
  expect(text.length).toEqual(1);
});