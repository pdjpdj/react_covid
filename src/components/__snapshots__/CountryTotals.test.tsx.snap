// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders country totals with data matches snapshot 1`] = `
<div>
  <table>
    <caption
      class="totals-subheader"
    >
      Latest (
      Thu 21 Jan 2021
      )
    </caption>
    <tbody>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Confirmed:
        </td>
        <td
          class="table-row-data"
        >
          4,500
        </td>
      </tr>
      <tr
        class="table-row table-row-alt"
      >
        <td
          class="table-row-label"
        >
          Deaths:
        </td>
        <td
          class="table-row-data"
        >
          133
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Recovered:
        </td>
        <td
          class="table-row-data"
        >
          340
        </td>
      </tr>
    </tbody>
  </table>
  <table>
    <caption
      class="totals-subheader"
    >
      All time
    </caption>
    <tbody>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Confirmed:
        </td>
        <td
          class="table-row-data"
        >
          9,000
        </td>
      </tr>
      <tr
        class="table-row table-row-alt"
      >
        <td
          class="table-row-label"
        >
          Deaths:
        </td>
        <td
          class="table-row-data"
        >
          256
        </td>
      </tr>
      <tr
        class="table-row"
      >
        <td
          class="table-row-label"
        >
          Recovered:
        </td>
        <td
          class="table-row-data"
        >
          680
        </td>
      </tr>
      <tr
        class="table-row table-row-alt"
      >
        <td
          class="table-row-label"
        >
          Active:
        </td>
        <td
          class="table-row-data"
        >
          8,320
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
